disperser/meterer/offchain_store.go

```diff
@@ -0,0 +1,269 @@
+	}
+
+	var reservation ReservationBin
+	err = attributevalue.UnmarshalMap(result, &reservation)
+	if err != nil {
+		return nil, err
+	}
+
+	return &reservation, nil
+}
+
+// Find all reservation bins for a given account
+func (s *OffchainStore) FindReservationBins(ctx context.Context, accountID string) ([]ReservationBin, error) {
+	result, err := s.dynamoClient.QueryIndex(ctx, s.reservationTableName, "AccountIDIndex", "AccountID = :accountID", commondynamodb.ExpresseionValues{
+		":accountID": &types.AttributeValueMemberS{Value: accountID},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if result == nil {
+		return nil, errors.New("reservation not found")
+	}
+
+	var reservations []ReservationBin
+	err = attributevalue.UnmarshalListOfMaps(result, &reservations)
+	if err != nil {
+		return nil, err
+	}
+
+	return reservations, nil
+}
+
+func (s *OffchainStore) AddOnDemandPayment(ctx context.Context, blobHeader BlobHeader) error {
+	result, err := s.dynamoClient.GetItem(ctx, s.onDemandTableName,
+		commondynamodb.Item{
+			"AccountID":          &types.AttributeValueMemberS{Value: blobHeader.AccountID},
+			"CumulativePayments": &types.AttributeValueMemberS{Value: strconv.FormatUint(blobHeader.CumulativePayment, 10)},
+		},
+	)
+	if err != nil {
+		fmt.Println("new payment record: %w", err)
+	}
+	if result != nil {
+		return fmt.Errorf("exact payment already exists")
+	}
+	err = s.dynamoClient.PutItem(ctx, s.onDemandTableName,
+		commondynamodb.Item{
+			"AccountID":          &types.AttributeValueMemberS{Value: blobHeader.AccountID},
+			"CumulativePayments": &types.AttributeValueMemberS{Value: strconv.FormatUint(blobHeader.CumulativePayment, 10)},
+			"BlobSize":           &types.AttributeValueMemberN{Value: strconv.FormatUint(uint64(blobHeader.BlobSize), 10)},
+		},
+	)
+
+	if err != nil {
+		return fmt.Errorf("failed to add payment: %w", err)
+	}
+	return nil
+}
+
+// RemoveOnDemandPayment removes a specific payment from the list for a specific account
+func (s *OffchainStore) RemoveOnDemandPayment(ctx context.Context, accountID string, payment uint64) error {
+	err := s.dynamoClient.DeleteItem(ctx, s.onDemandTableName,
+		commondynamodb.Key{
+			"AccountID":          &types.AttributeValueMemberS{Value: accountID},
+			"CumulativePayments": &types.AttributeValueMemberS{Value: strconv.FormatUint(payment, 10)},
+		},
+	)
+
+	if err != nil {
+		return fmt.Errorf("failed to remove payment: %w", err)
+	}
+
+	return nil
+}
+
+// relevant on-demand payment records: previous cumulative payment, next cumulative payment, blob size of next payment
+func (s *OffchainStore) GetRelevantOnDemandRecords(ctx context.Context, accountID string, cumulativePayment uint64) (uint64, uint64, uint32, error) {
+	result, err := s.dynamoClient.QueryIndex(ctx, s.onDemandTableName, "AccountIDIndex", "AccountID = :account", commondynamodb.ExpresseionValues{
+		":account": &types.AttributeValueMemberS{
+			Value: accountID,
+		}})
```

Comment:
I think we should be able to configure a sort key associated with the `CumulativePayment` field and directly retrieve the records with cumulative payment above and below the provided payment.

Reply:
updated!